core/src/cas/file/actor.rs

```diff
@@ -0,0 +1,131 @@
+use actor::{Protocol, SYS};
+use cas::content::{Address, Content};
+use error::HolochainError;
+use riker::actors::*;
+use std::{
+    fs::{create_dir_all, read_to_string, write},
+    path::{Path, MAIN_SEPARATOR},
+};
+
+const ACTOR_ID_ROOT: &'static str = "filesystem_storage_actor";
+
+fn path_to_actor_id(dir_path: &str) -> String {
+    format!("{}{}{}", ACTOR_ID_ROOT, MAIN_SEPARATOR, dir_path)
+}
+
+pub struct FilesystemStorageActor {
+    /// path to the directory where content will be saved to disk
+    dir_path: String,
+}
+
+impl FilesystemStorageActor {
+    pub fn new(dir_path: String) -> FilesystemStorageActor {
+        FilesystemStorageActor { dir_path }
+    }
+
+    /// actor() for riker
+    fn actor(dir_path: String) -> BoxActor<Protocol> {
+        Box::new(FilesystemStorageActor::new(dir_path))
+    }
+
+    /// props() for riker
+    fn props(dir_path: &str) -> BoxActorProd<Protocol> {
+        Props::new_args(
+            Box::new(FilesystemStorageActor::actor),
+            dir_path.to_string(),
+        )
+    }
+
+    pub fn new_ref(dir_path: &str) -> Result<ActorRef<Protocol>, HolochainError> {
+        let canonical = Path::new(&dir_path).canonicalize()?;
+        if !canonical.is_dir() {
+            return Err(HolochainError::IoError(
+                "path is not a directory or permissions don't allow access".to_string(),
+            ));
+        }
+        let dir_path = canonical
+            .to_str()
+            .ok_or_else(|| HolochainError::IoError("could not convert path to string".to_string()))?
+            .to_string();
+        Ok(SYS.actor_of(
+            FilesystemStorageActor::props(&dir_path),
+            // always return the same reference to the same actor for the same path
+            // consistency here provides safety for CAS methods
+            &path_to_actor_id(&dir_path),
+        )?)
+    }
+
+    /// builds an absolute path for an AddressableContent address
+    fn address_to_path(&self, address: &Address) -> String {
+        // using .txt extension because content is arbitrary and controlled by the
+        // AddressableContent trait implementation
+        format!("{}{}{}.txt", self.dir_path, MAIN_SEPARATOR, address)
+    }
+
+    /// filesystem CAS add. NOT thread safe.
+    fn unsafe_add(&self, address: &Address, content: &Content) -> Result<(), HolochainError> {
+        // @TODO be more efficient here
+        // @see https://github.com/holochain/holochain-rust/issues/248
+        create_dir_all(&self.dir_path)?;
+        Ok(write(self.address_to_path(address), content)?)
+    }
+
+    /// filesystem CAS contains. NOT thread safe.
+    fn unsafe_contains(&self, address: &Address) -> Result<bool, HolochainError> {
+        Ok(Path::new(&self.address_to_path(address)).is_file())
+    }
+
+    /// filesystem CAS fetch. NOT thread safe.
+    fn unsafe_fetch(&self, address: &Address) -> Result<Option<Content>, HolochainError> {
```

Comment:
I was afraid we were using unsafe rust and my eyes popped out haha. I think the naming gives the impression that we are using a subsection of rust that behaves in an unsafe context https://doc.rust-lang.org/book/second-edition/ch19-01-unsafe-rust.html

Reply:
@AshantiMutinta that's fair. would you accept `unthreadable_fetch`?

Reply:
unthreadable is good but then again we have to ask what properties does this have which make them so unthreadable. I guess it is the lack of the ARC for the context? In that case I think that is a good name. Just highlighting the question of, if I had to put this in a thread::spawn, what would stop me? Is it unthreadable or unthreaded?

Reply:
Random suggestion `single_threaded_fetch()` or  `st_fetch()`